pkg/kubelet/kubelet.go

```diff
@@ -2665,49 +2633,28 @@ func (kl *Kubelet) HandlePodAdditions(pods []*v1.Pod) {
 			// We failed pods that we rejected, so allocatedPods include all admitted
 			// pods that are alive.
 			allocatedPods := kl.getAllocatedPods()
-			// Filter out the pod being evaluated.
-			allocatedPods = slices.DeleteFunc(allocatedPods, func(p *v1.Pod) bool { return p.UID == pod.UID })
-
-			if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
-				// To handle kubelet restarts, test pod admissibility using AllocatedResources values
-				// (for cpu & memory) from checkpoint store. If found, that is the source of truth.
-				allocatedPod, _ := kl.allocationManager.UpdatePodFromAllocation(pod)
-
-				// Check if we can admit the pod; if not, reject it.
-				if ok, reason, message := kl.canAdmitPod(allocatedPods, allocatedPod); !ok {
-					kl.rejectPod(pod, reason, message)
-					// We avoid recording the metric in canAdmitPod because it's called
-					// repeatedly during a resize, which would inflate the metric.
-					// Instead, we record the metric here in HandlePodAdditions for new pods
-					// and capture resize events separately.
-					recordAdmissionRejection(reason)
-					continue
-				}
-				// For new pod, checkpoint the resource values at which the Pod has been admitted
-				if err := kl.allocationManager.SetAllocatedResources(allocatedPod); err != nil {
-					//TODO(vinaykul,InPlacePodVerticalScaling): Can we recover from this in some way? Investigate
-					klog.ErrorS(err, "SetPodAllocation failed", "pod", klog.KObj(pod))
-				}
-			} else {
-				// Check if we can admit the pod; if not, reject it.
-				if ok, reason, message := kl.canAdmitPod(allocatedPods, pod); !ok {
-					kl.rejectPod(pod, reason, message)
-					// We avoid recording the metric in canAdmitPod because it's called
-					// repeatedly during a resize, which would inflate the metric.
-					// Instead, we record the metric here in HandlePodAdditions for new pods
-					// and capture resize events separately.
-					recordAdmissionRejection(reason)
-					continue
-				}
+
+			// Check if we can admit the pod; if not, reject it.
+			if ok, reason, message := kl.allocationManager.AddPod(allocatedPods, pod); !ok {
+				kl.rejectPod(pod, reason, message)
+				// We avoid recording the metric in canAdmitPod because it's called
+				// repeatedly during a resize, which would inflate the metric.
+				// Instead, we record the metric here in HandlePodAdditions for new pods
+				// and capture resize events separately.
+				recordAdmissionRejection(reason)
+				continue
 			}
 		}
-		kl.podWorkers.UpdatePod(UpdatePodOptions{
+		updates = append(updates, UpdatePodOptions{
 			Pod:        pod,
 			MirrorPod:  mirrorPod,
 			UpdateType: kubetypes.SyncPodCreate,
 			StartTime:  start,
 		})
 	}
+	for _, update := range updates {
```

Comment:
nit: Add a comment about why UpdatePod needs to be called after all the pods are admitted.